controllers/clustertopology_controller_compute.go

```diff
@@ -78,8 +84,286 @@ func (r *ClusterTopologyReconciler) getCurrentState(ctx context.Context, cluster
+// subset of a topology will be implemented.
+func (r *ClusterTopologyReconciler) computeDesiredState(_ context.Context, class *clusterTopologyClass, current *clusterTopologyState) (*clusterTopologyState, error) {
+	var err error
+	desiredState := &clusterTopologyState{}
+
+	// Compute the desired state of the InfrastructureCluster object.
+	if desiredState.infrastructureCluster, err = computeInfrastructureCluster(class, current); err != nil {
+		return nil, err
+	}
+
+	// If the control plane object requires it, compute the InfrastructureMachineTemplate for the control plane.
+	if class.controlPlane.infrastructureMachineTemplate != nil {
+		desiredState.controlPlane.infrastructureMachineTemplate = computeControlPlaneInfrastructureMachineTemplate(class, current)
+	}
+
+	// Compute the desired state of the ControlPlane object, eventually adding a reference to the
+	// InfrastructureMachineTemplate generated by the previous step.
+	if desiredState.controlPlane.object, err = computeControlPlane(class, current, desiredState.controlPlane.infrastructureMachineTemplate); err != nil {
+		return nil, err
+	}
+
+	// Compute the desired state for the cluster object adding a reference to the
+	// InfrastructureCluster and the ControlPlane objects generated by the previous step.
+	desiredState.cluster = computeCluster(current, desiredState.infrastructureCluster, desiredState.controlPlane.object)
+
+	// TODO: implement generate desired state for machine deployments
+
+	return desiredState, nil
+}
+
+// computeInfrastructureCluster computes the desired state for the InfrastructureCluster object starting from the
+// corresponding template defined in cluster class.
+func computeInfrastructureCluster(class *clusterTopologyClass, current *clusterTopologyState) (*unstructured.Unstructured, error) {
+	infrastructureCluster, err := templateToObject(templateToInput{
+		template:              class.infrastructureClusterTemplate,
+		templateClonedFromRef: class.clusterClass.Spec.Infrastructure.Ref,
+		cluster:               current.cluster,
+		namePrefix:            fmt.Sprintf("%s-", current.cluster.Name),
+		currentObjectRef:      current.cluster.Spec.InfrastructureRef,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to generate the cluster infrastructure object")
+	}
+	return infrastructureCluster, nil
+}
+
+// computeControlPlaneInfrastructureMachineTemplate computes the desired state for InfrastructureMachineTemplate
+// that should be referenced by the control plane object.
+func computeControlPlaneInfrastructureMachineTemplate(class *clusterTopologyClass, current *clusterTopologyState) *unstructured.Unstructured {
+	controlPlaneInfrastructureMachineTemplate := templateToTemplate(templateToInput{
+		template:              class.controlPlane.infrastructureMachineTemplate,
+		templateClonedFromRef: objToRef(class.controlPlane.infrastructureMachineTemplate),
```

Comment:
```suggestion
		templateClonedFromRef: class.clusterClass.Spec.ControlPlane.MachineInfrastructure.Ref,
```
after #5000, right? (mostly to confirm my understanding)

Reply:
yes!